### PR TITLE
Add agit flow support in gitea

### PR DESCRIPTION
**Pull request by a1012112796**
ref: https://git-repo.info/en/2020/03/agit-flow-and-git-repo/

examle:
test

NOTICE: All existed repositories will not have the new hooks, you have to rewrite the hooks via admin panel.

### CODE REVIEW

**Comment by lunny** [comment]
@jiangxin Could we ask you to help to review this PR?

File: models/pull.go [hunk 1]
```diff
@@ -57,6 +69,7 @@ type PullRequest struct {
 	BaseRepo        *Repository `xorm:"-"`
 	HeadBranch      string
 	BaseBranch      string
+	TopicBranch     string           // use for agit style pr
```
Comment (lunny) [inline models/pull.go]: Could we still use column `HeadBranch` since we have `Style` column, we can know `HeadBranch` is `TopicBranch`?

File: models/pull.go [hunk 2]
```diff
@@ -467,8 +482,8 @@ func newPullRequestAttempt(repo *Repository, pull *Issue, labelIDs []int64, uuid
 func GetUnmergedPullRequest(headRepoID, baseRepoID int64, headBranch, baseBranch string) (*PullRequest, error) {
```
Comment (lunny) [inline models/pull.go]: We can give a style parameter.

File: cmd/hook.go [hunk 3]
```diff
@@ -460,3 +473,261 @@ func pushOptions() map[string]string {
...
+	}
+
+	r.Type = pktLineTypeData
+	r.Data = string(tmp)
+
+	return r, nil
+}
+
+func writePktLine(out io.Writer, typ pktLineType, data []byte) error {
+	if typ == pktLineTypeFlush {
```
Comment (jiangxin) [inline cmd/hook.go]: You can define two different functions, one write pkt-line, and the other write flush packet.

File: cmd/hook.go [hunk 4]
```diff
@@ -460,3 +473,261 @@ func pushOptions() map[string]string {
...
+		if rs.Type == pktLineTypeFlush {
+			break
+		}
+		t := strings.SplitN(rs.Data, " ", 3)
+		if len(t) != 3 {
+			continue
+		}
+		hookOptions.OldCommitIDs = append(hookOptions.OldCommitIDs, t[0])
+		hookOptions.NewCommitIDs = append(hookOptions.NewCommitIDs, t[1])
+		hookOptions.RefFullNames = append(hookOptions.RefFullNames, t[2])
```
Comment (jiangxin) [inline cmd/hook.go:605]: It's better to store the 3 variables in one struct such as "Command", and use a slice of Commands to hold all commands from user.

File: cmd/hook.go [hunk 5]
```diff
@@ -460,3 +473,261 @@ func pushOptions() map[string]string {
...
+		for {
+			rs, err = readPktLine(reader)
+			if err != nil {
+				fail("Internal Server Error", "Pkt-Line format is wrong :%v", err)
+			}
+			if rs.Type == pktLineTypeFlush {
+				break
+			}
+
+			kv := strings.SplitN(rs.Data, "=", 2)
```
Comment (jiangxin) [inline cmd/hook.go]: "push-option" is still valid if there is no equal sign in it, it's safe to ignore it in this case.

File: cmd/hook.go [hunk 6]
```diff
@@ -460,3 +473,261 @@ func pushOptions() map[string]string {
...
+	// # OK, but has an alternate reference.  The alternate reference name
+	// # and other status can be given in option directives.
+	// H: PKT-LINE(ok <ref>)
+	// H: PKT-LINE(option refname <refname>)
+	// H: PKT-LINE(option old-oid <old-oid>)
+	// H: PKT-LINE(option new-oid <new-oid>)
+	// H: PKT-LINE(option forced-update)
+	// H: ... ...
+	// H: flush-pkt
+	for _, rs := range resp.Results {
```
Comment (jiangxin) [inline cmd/hook.go:650]: There are 4 types of responses supported from hook to receive-pack, but you only implement one of them.

1. OK, run this command successfully.

        H: PKT-LINE(ok <ref>)

2. NO, I reject it.

        H: PKT-LINE(ng <ref> <reason>)

3. Fall through, let 'receive-pack' to execute it.

        H: PKT-LINE(ok <ref>)
        H: PKT-LINE(option fall-through)

4. OK, but has an alternate reference.  The alternate reference name and other status can be given in option directives.

        H: PKT-LINE(ok <ref>)
        H: PKT-LINE(option refname <refname>)
        H: PKT-LINE(option old-oid <old-oid>)
        H: PKT-LINE(option new-oid <new-oid>)
        H: PKT-LINE(option forced-update)

File: cmd/hook.go [hunk 7]
```diff
@@ -460,3 +473,261 @@ func pushOptions() map[string]string {
...
+		err = writePktLine(os.Stdout, pktLineTypeData, []byte("ok "+rs.OrignRef))
+		if err != nil {
+			fail("Internal Server Error", "Pkt-Line response failed: %v", err)
+		}
+		err = writePktLine(os.Stdout, pktLineTypeData, []byte("option refname "+rs.Ref))
+		if err != nil {
+			fail("Internal Server Error", "Pkt-Line response failed: %v", err)
+		}
+		err = writePktLine(os.Stdout, pktLineTypeData, []byte("option old-oid "+rs.OldOID))
```
Comment (jiangxin) [inline cmd/hook.go]: if `rs.OldOID` is empty, ignore it.

File: cmd/hook.go [hunk 8]
```diff
@@ -460,3 +473,261 @@ func pushOptions() map[string]string {
...
+	// # and other status can be given in option directives.
+	// H: PKT-LINE(ok <ref>)
+	// H: PKT-LINE(option refname <refname>)
+	// H: PKT-LINE(option old-oid <old-oid>)
+	// H: PKT-LINE(option new-oid <new-oid>)
+	// H: PKT-LINE(option forced-update)
+	// H: ... ...
+	// H: flush-pkt
+	for _, rs := range resp.Results {
+		err = writePktLine(os.Stdout, pktLineTypeData, []byte("ok "+rs.OrignRef))
```
Comment (jiangxin) [inline cmd/hook.go]: OK, this one must be present.

File: cmd/hook.go [hunk 9]
```diff
@@ -460,3 +473,261 @@ func pushOptions() map[string]string {
...
+		if err != nil {
+			fail("Internal Server Error", "Pkt-Line response failed: %v", err)
+		}
+		err = writePktLine(os.Stdout, pktLineTypeData, []byte("option refname "+rs.Ref))
```
Comment (jiangxin) [inline cmd/hook.go]: This one can be omitted, if `rs.Ref` is NULL.

File: cmd/hook.go [hunk 10]
```diff
@@ -460,3 +473,261 @@ func pushOptions() map[string]string {
...
+	if err != nil {
+		fail("Internal Server Error", "Pkt-Line format is wrong :%v", err)
+	}
+	if rs.Type != pktLineTypeData {
+		fail("Internal Server Error", "Pkt-Line format is wrong. get %v", rs)
+	}
+
+	const VersionHead string = "version=1"
+
+	if !strings.HasPrefix(rs.Data, VersionHead) {
```
Comment (jiangxin) [inline cmd/hook.go]: You should split `rs.Data` by the NUL character ('\0').  Part one is version, and part two is capabilities.

File: cmd/hook.go [hunk 11]
```diff
@@ -460,3 +473,261 @@ func pushOptions() map[string]string {
...
+		fail("Internal Server Error", "Pkt-Line format is wrong. get %v", rs)
+	}
+
+	hasPushOptions := false
+	response := []byte(VersionHead)
+	if strings.Contains(rs.Data, "push-options") {
```
Comment (jiangxin) [inline cmd/hook.go]: You should split capabilities using space character, and try to match each capability.
If there is a new capability called "push-options-v2", using `string.Contains` is wrong.

File: cmd/hook.go [hunk 12]
```diff
@@ -460,3 +473,261 @@ func pushOptions() map[string]string {
...
+		response = append(response, byte(0))
+		response = append(response, []byte("push-options")...)
+		hasPushOptions = true
+	}
+	response = append(response, []byte("\n")...)
```
Comment (jiangxin) [inline cmd/hook.go]: Append a `LF` is good.

File: cmd/hook.go [hunk 13]
```diff
@@ -460,3 +473,261 @@ func pushOptions() map[string]string {
...
+	// H: flush-pkt
+
+	rs, err := readPktLine(reader)
+	if err != nil {
+		fail("Internal Server Error", "Pkt-Line format is wrong :%v", err)
+	}
+	if rs.Type != pktLineTypeData {
+		fail("Internal Server Error", "Pkt-Line format is wrong. get %v", rs)
+	}
+
```
Comment (jiangxin) [inline cmd/hook.go:522]: There are many duplicate code to parse reader and handle error output. It's better to parse reader and send response in one loop.

**Review by jiangxin — CHANGES_REQUESTED** [review]

**Comment by jiangxin** [comment]
It's not good to have a merge commit for others to review, so please rebase your topic.

**Comment by a1012112796** [comment]
@jiangxin Thanks for your carefully check, I will change them soon.

File: cmd/hook.go (hunk 6 above)
Comment (a1012112796) [reply cmd/hook.go:650]: I know it, but other types is not usefull now.
@jiangxin By the way, I have a small question, has any way to check if user push with ``-f`` option in this hook? Thanks.

File: cmd/hook.go [hunk 14]
```diff
@@ -460,3 +473,261 @@ func pushOptions() map[string]string {
...
+	// flush-pkt "0000"
+	pktLineTypeFlush pktLineType = iota
+	// data line
+	pktLineTypeData
+)
+
+// gitPktLine pkt-line api
+type gitPktLine struct {
+	Type   pktLineType
+	Length int64
```
[the remark this answers is not in the document]
Comment (a1012112796) [reply cmd/hook.go]: how about uint64? Thanks
which used in many go std libs.

File: cmd/hook.go [hunk 15]
```diff
@@ -460,3 +473,261 @@ func pushOptions() map[string]string {
...
+	}
+
+	if r.Length <= 4 || r.Length > 65520 {
+		r.Type = pktLineTypeUnknow
+		return r, nil
+	}
+
+	tmp := make([]byte, r.Length-4)
+	for i := range tmp {
+		tmp[i], err = in.ReadByte()
```
[the remark this answers is not in the document]
Comment (a1012112796) [reply cmd/hook.go]: Don't warry, I have use ``bufio.Reader``, which will do what you like in it.

File: cmd/hook.go (hunk 4 above)
Comment (a1012112796) [reply cmd/hook.go:605]: It's ok, but this struct has been used in other hooks, change it will broken other code also ...
In my opion, It not a big problem.

File: cmd/hook.go (hunk 9 above)
Comment (a1012112796) [reply cmd/hook.go]: In current design, It will't be empty.

**Comment by a1012112796** [comment]
@jiangxin Another question, I know your designed style to create pull request is ``git push origin HEAD:refs/for/<target-branch>/<session>``, But the ``<target-branch>`` also maybe contain ``/``.
I wonder.... (now I have to change it to ``git push origin HEAD:refs/for/<target-branch> -o topic=<session>``).
Can you tell me reason? Thanks.

**Comment by jiangxin** [comment]
> @jiangxin Another question, I know your designed style to create pull request is `git push origin HEAD:refs/for/<target-branch>/<session>`, But the `<target-branch>` also maybe contain `/`.
> I wonder.... (now I have to change it to `git push origin HEAD:refs/for/<target-branch> -o topic=<session>`).
> Can you tell me reason? Thanks.

Gerrit has the same syntax, see: https://github.com/alibaba/git-repo-go/blob/master/helper/proto-gerrit.go#L105-L135
Even using slash to separate branch name and topic, it's not difficult to find the proper branch name. With the help of `proc-receive` hook, we can have our own syntax of cause, but follow the syntax of Gerrit is better.

**Comment by a1012112796** [comment]
> > @jiangxin Another question, I know your designed style to create pull request is `git push origin HEAD:refs/for/<target-branch>/<session>`, But the `<target-branch>` also maybe contain `/`.
> > I wonder.... (now I have to change it to `git push origin HEAD:refs/for/<target-branch> -o topic=<session>`).
> > Can you tell me reason? Thanks.
> 
> Gerrit has the same syntax, see: https://github.com/alibaba/git-repo-go/blob/master/helper/proto-gerrit.go#L105-L135
> Even using slash to separate branch name and topic, it's not difficult to find the proper branch name. With the help of `proc-receive` hook, we can have our own syntax of cause, but follow the syntax of Gerrit is better.

but I don't know how to split them... 

for example
``refs/for/release/v1.12/bugfix/xxxx``

**Comment by jiangxin** [comment]
> but I don't know how to split them...
> 
> for example
> `refs/for/release/v1.12/bugfix/xxxx`

Try to validate each of the following references in order, and the first valid reference is the target branch:

+ refs/heads/release
+ refs/heads/release/v1.12
+ refs/heads/release/v1.12/bugfix
+ refs/heads/release/v1.12/bugfix/xxxx

**Comment by a1012112796** [comment]
> > but I don't know how to split them...
> > for example
> > `refs/for/release/v1.12/bugfix/xxxx`
> 
> Try to validate each of the following references in order, and the first valid reference is the target branch:
> 
> * refs/heads/release
> * refs/heads/release/v1.12
> * refs/heads/release/v1.12/bugfix
> * refs/heads/release/v1.12/bugfix/xxxx

Okay, Thanks. I see, but that means this logic maybe make wrong split. (for example: if branch ``release`` and ``release\v1.12`` all exist). so I'd like not only suport ``refs/for/<target-branch>/<topic-branch>`` but also ``refs/for/<target-branch> -o topic='<topic-branch>'``. :)

**Comment by lafriks** [comment]
> Okay, Thanks. I see, but that means this logic maybe make wrong split. (for example: if branch `release` and `release/v1.12` all exist). so I'd like not only suport `refs/for/<target-branch>/<topic-branch>` but also `refs/for/<target-branch> -o topic='<topic-branch>'`. :)

You can not have branch `release` and branch `release/v1.12` at the same time, git does not allow that

File: .drone.yml [hunk 16]
```diff
@@ -231,10 +231,8 @@ steps:
         from_secret: github_read_token
 
   - name: test-mysql
-    image: golang:1.15
+    image: a1012112796/gitea_test_env:linux-amd64
```
Comment (lafriks) [inline .drone.yml]: This is unrelated change for this PR

Comment (a1012112796) [reply .drone.yml]: Without this change, can't run new test code for this feature because of the low git version.

File: cmd/hook.go [hunk 17]
```diff
@@ -460,3 +473,256 @@ func pushOptions() map[string]string {
...
+	}
+
+	for _, option := range requestOptions {
+		if strings.HasPrefix(option, "push-options") {
+			response = append(response, byte(0))
+			response = append(response, []byte("push-options")...)
+			hasPushOptions = true
+		}
+	}
+	response = append(response, []byte("\n")...)
```
Comment (lunny) [inline cmd/hook.go]: Just '\n'

File: models/pull.go [hunk 18]
```diff
@@ -58,6 +68,7 @@ type PullRequest struct {
 	BaseRepoID      int64       `xorm:"INDEX"`
 	BaseRepo        *Repository `xorm:"-"`
 	HeadBranch      string
+	HeadCommitID    string `xorm:"-"`
```
Comment (lunny) [inline models/pull.go:71]: Maybe `HeadRefName` is a better name?

File: modules/git/repo_branch.go [hunk 19]
```diff
@@ -32,6 +40,19 @@ type Branch struct {
 	gitRepo *Repository
 }
 
+// GetRefCommitID returns the last commit ID string of given reference (branch or tag).
+func GetRefCommitID(repoPath, name string) (string, error) {
```
Comment (lunny) [inline modules/git/repo_branch.go]: I think this has been implemented somewhere.

File: models/pull.go (hunk 18 above)
Comment (a1012112796) [reply models/pull.go:71]: I think ``HeadCommitID``  is okay becauce it only used to save a commit id when a agit style pull request is not created (so can't get it by pull head ref).

**Comment by lafriks** [comment]
Please resolve conflicts

**Comment by zeripath** [comment]
conflicts resolved.

File: cmd/hook.go (hunk 4 above)
Comment (zeripath) [reply cmd/hook.go:605]: yeah this is fine. It's not a Command slice like @jiangjianxiao suggests as we would only have to immediately decompose that slice into the lists that get sent to the server.

File: cmd/hook.go (hunk 6 above)
Comment (zeripath) [reply cmd/hook.go:650]: I think we need to ensure that:

```
H: PKT-LINE(ok <ref>) 
H: PKT-LINE(option fall-through)
```

is sent if we're sent a reference that isn't matching the ones we expect.

**Comment by lafriks** [comment]
does not seems to be resolved :)

File: modules/private/hook.go [hunk 20]
```diff
@@ -79,6 +81,22 @@ type HookPostReceiveBranchResult struct {
...
+	Results []HockProcReceiveRefResult
+	Err     string
+}
+
+// HockProcReceiveRefResult represents an individual result from ProcReceive
+type HockProcReceiveRefResult struct {
+	OldOID      string
+	NewOID      string
+	Ref         string
+	OrignRef    string
```
Comment (zeripath) [inline modules/private/hook.go]: I don't like this OrignRef name - This is the original ref that was provided in the proc-receive so either:

```suggestion
	OriginalRef    string
```

or 

```suggestion
	ProvidedRef    string
```

We can probably leave the Ref one as is, but it could be called `AlternateRef`

**Comment by 6543** [comment]
if you exec push command twice: a new pull get created and now we have a duplicate !

**Comment by 6543** [comment]
I could create a broken pull by creating a new pull `git push origin HEAD:refs/for/main -o topic=coffee` and target it second: `git push origin HEAD:refs/for/refs/pull/3/head -o topic=coffee`

**Comment by 6543** [comment]
**this does not work** as per `https://git-repo.info/en/2020/03/agit-flow-and-git-repo/` "AGit-Flow Implementation":
 * “refs/drafts/”,  -> add a WIP: as title prefix (dont hardcode wip use the first of the draft prefix set via settings)
 * “refs/for-review/”  -> no idear just create a normal pull as `refs/for` do?

## important: now a read only user can insert infalid refs as he like just test `git push origin HEAD:refs/nonesence/main -o topic=coffee`, this wont show up but get created!

**Comment by zeripath** [comment]
Agh. That's not good!

**Comment by lunny** [comment]
> ### important: now a read only user can insert infalid refs as he like just test `git push origin HEAD:refs/nonesence/main -o topic=coffee`, this wont show up but get created!

PR writer permission should be checked here.

**Comment by a1012112796** [comment]
> “refs/drafts/”, -> add a WIP: as title prefix (dont hardcode wip use the first of the draft prefix set via settings)

Not so necessary, because we not have a special mode named as `draft` like gh, user can do it
by change title manually.
> “refs/for-review/” -> no idear just create a normal pull as refs/for do?

Need more work, not like do it in this pull request.
two choces
- create a speciall review. (ref: #10007)
- create a special pull request (the target branch is a special ref), 

>  important: now a read only user can insert infalid refs as he like just test git push origin HEAD:refs/nonesence/main -o topic=coffee, this wont show up but get created!

I see, It's bug, will do something to prevent it.

> ``git push origin HEAD:refs/for/refs/pull/3/head -o topic=coffee``

will do mre work for it. now just prevent it.

**Comment by a1012112796** [comment]
> if you exec push command twice: a new pull get created and now we have a duplicate !

@6543 Hasn't found ths bug, see test code.

**Comment by 6543** [comment]
with agit created pulls dont show up in `http://gitew:3000/pulls?type=created_by&repos=[]&sort=&state=open`

**Comment by 6543** [comment]
unique constrain for creating a new pull still not completly enforced:
Do:
1. `git push origin HEAD:refs/for/main -o topic=coffee_no_milk`
2. `git push origin HEAD:refs/for/main -o topic=coffee_with_milk`

**Comment by a1012112796** [comment]
> unique constrain for creating a new pull still not completly enforced:
> Do:
> 
> 1. `git push origin HEAD:refs/for/main -o topic=coffee_no_milk`
> 2. `git push origin HEAD:refs/for/main -o topic=coffee_with_milk`

image

They are different, Aren't them?

**Comment by 6543** [comment]
well if topic is is the only unique constrain factor ... yes
since you dont have a "base-branch + head-branch" relation but just a "base-branch head ref" where ref is topic related ... it should be fine, it's just bit strange :/

**Comment by a1012112796** [comment]
> well if topic is is the only unique constrain factor ... yes
> since you dont have a "base-branch + head-branch" relation but just a "base-branch head ref" where ref is topic related ... it should be fine, it's just bit strange :/


In my opion, the ``topic`` still can be  seen as head branch, the different is that not create a real git branch, but record it in the database(maybe like draft release). when push again will find pull id from database by ``topic``, then update pull ref (``refs/pull/xxx/head``) directly.

File: routers/private/hook.go [hunk 21]
```diff
@@ -574,6 +635,278 @@ func HookPostReceive(ctx *gitea_context.PrivateContext) {
...
+			ctx.JSON(http.StatusInternalServerError, map[string]interface{}{
+				"Err": fmt.Sprintf("Failed to load pull issue. Error: %v", err),
+			})
+			return
+		}
+		comment, err := models.CreatePushPullComment(pusher, pr, oldCommitID, opts.NewCommitIDs[i])
+		if err == nil && comment != nil {
+			notification.NotifyPullRequestPushCommits(pusher, pr, comment)
+		}
+		notification.NotifyPullRequestSynchronized(pusher, pr)
```
Comment (6543) [inline routers/private/hook.go]: I think we should not put the whole aGit flow into hook module

Can we move it into a dedicated module at services?

**Comment by 6543** [comment]
I would propose `services/agit/x...go`

File: modules/git/repo_branch.go [hunk 22]
```diff
@@ -13,6 +13,14 @@ import (
 // BranchPrefix base dir of the branch information file store on git
 const BranchPrefix = "refs/heads/"
 
+// AGit Flow
+
+// PullRequestPrefix sepcial ref to create a pull request: refs/for/<targe-branch>/<topic-branch>
+// or refs/for/<targe-branch> -o topic='<topic-branch>'
+const PullRequestPrefix = "refs/for/"
```
Comment (6543) [inline modules/git/repo_branch.go:20]: can you move this const into **services/agit/agit.go** ?

Comment (6543) [reply modules/git/repo_branch.go:20]: never mind - just checked how other refs prefix are handled and it need a refactor ...
-> I propose to move things into modules/git/ref.go and replace strings in code with that const ... witch is worth it's own pull

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
one thing but this can be done in a follow up pull: let the repo admin decide if he like to accept agit pulls or not

File: cmd/hook.go [hunk 23]
```diff
@@ -463,3 +483,317 @@ func pushOptions() map[string]string {
...
+	const VersionHead string = "version=1"
+
+	hasPushOptions := false
+	response := []byte(VersionHead)
+	var requestOptions []string
+
+	for i := range rs.Data {
+		if rs.Data[i] == byte(0) {
+			if string(rs.Data[0:i]) != VersionHead {
+				return fail("Internal Server Error", "Received an not supported version: %s", string(rs.Data[0:i]))
```
Comment (lunny) [inline cmd/hook.go]: unsupported

File: cmd/hook.go [hunk 24]
```diff
@@ -463,3 +483,317 @@ func pushOptions() map[string]string {
...
+	rs, err := readPktLine(reader, pktLineTypeData)
+	if err != nil {
+		return err
+	}
+
+	const VersionHead string = "version=1"
+
+	hasPushOptions := false
+	response := []byte(VersionHead)
+	var requestOptions []string
```
Comment (lunny) [inline cmd/hook.go]: It's better to use `var` for the three definitions.

File: cmd/hook.go [hunk 25]
```diff
@@ -463,3 +483,317 @@ func pushOptions() map[string]string {
...
+
+	for i := range rs.Data {
```
Comment (lunny) [inline cmd/hook.go]: It's better to use `bytes.IndexByte`

File: models/pull.go [hunk 26]
```diff
@@ -38,6 +38,16 @@ const (
 	PullRequestStatusEmpty
 )
 
+// PullRequestStyle the style of pull request
+type PullRequestStyle int
```
Comment (lunny) [inline models/pull.go]: How about use `PullRequestFlow`?

File: modules/convert/pull.go [hunk 27]
```diff
@@ -95,7 +95,25 @@ func ToAPIPullRequest(pr *models.PullRequest) *api.PullRequest {
 		}
 	}
 
-	if pr.HeadRepo != nil {
+	if pr.Style != models.PullRequestStyleGithub {
```
Comment (lunny) [inline modules/convert/pull.go]: == PullRequestStyleAgit?

File: routers/private/hook.go [hunk 28]
```diff
@@ -155,6 +156,42 @@ func HookPreReceive(ctx *gitea_context.PrivateContext) {
...
+				}
+
+				perm, err := models.GetUserRepoPermission(repo, pusher)
+				if err != nil {
+					log.Error("models.GetUserRepoPermission:%v", err)
+					ctx.Error(http.StatusInternalServerError, "")
+					return
+				}
+
+				if !perm.CanWrite(models.UnitTypeCode) {
```
Comment (lunny) [inline routers/private/hook.go:181]: Maybe we should check if he has write permission to pull request but not to the code directly.

File: services/agit/agit.go [hunk 29]
```diff
@@ -0,0 +1,265 @@
...
+
+		headBranch := ""
+		userName := strings.ToLower(opts.UserName)
+
+		if len(curentTopicBranch) == 0 {
+			curentTopicBranch = topicBranch
+		}
+
+		if !strings.HasPrefix(curentTopicBranch, userName+"/") {
+			headBranch = userName + "/" + curentTopicBranch
```
Comment (lunny) [inline services/agit/agit.go:94]: Maybe we should use userid but not userName because username maybe deleted and recreated. Or we have to delete all these refs files when user is deleted.

File: services/pull/commit_status.go [hunk 30]
```diff
@@ -108,13 +108,21 @@ func GetPullRequestCommitStatusState(pr *models.PullRequest) (structs.CommitStat
...
-	if !headGitRepo.IsBranchExist(pr.HeadBranch) {
+	if pr.Style == models.PullRequestStyleGithub && !headGitRepo.IsBranchExist(pr.HeadBranch) {
+		return "", errors.New("Head branch does not exist, can not merge")
+	}
+	if pr.Style != models.PullRequestStyleGithub && !git.IsReferenceExist(headGitRepo.Path, pr.GetGitRefName()) {
 		return "", errors.New("Head branch does not exist, can not merge")
 	}
 
-	sha, err := headGitRepo.GetBranchCommitID(pr.HeadBranch)
+	sha := ""
```
Comment (lunny) [inline services/pull/commit_status.go]: var sha string

File: modules/convert/pull.go (hunk 27 above)
Comment (a1012112796) [reply modules/convert/pull.go]: They are same now, aren't them?

File: routers/private/hook.go (hunk 28 above)
Comment (a1012112796) [reply routers/private/hook.go:181]: This check is delayed permission check when user want to update other ref (banch, tag ...). so write permission to code is enough.

File: services/agit/agit.go (hunk 29 above)
Comment (a1012112796) [reply services/agit/agit.go:94]: I think it's not a problem, Because it will't be used to create any git refs. just use to checkout the exist unmerged pull request which has same topic and user name. (see: https://github.com/go-gitea/gitea/pull/14295#issuecomment-869100765)
after pull request closed or merged. This label willn't be used in anywhere.

File: models/migrations/v189.go [hunk 31]
```diff
@@ -11,10 +11,10 @@ import (
 )
 
 func addAgitStylePullRequest(x *xorm.Engine) error {
-	type PullRequestStyle int
+	type PullRequestFlow int
 
 	type PullRequest struct {
-		Style PullRequestStyle `xorm:"NOT NULL DEFAULT 0"`
+		Style PullRequestFlow `xorm:"NOT NULL DEFAULT 0"`
```
Comment (lunny) [inline models/migrations/v189.go]: Could we rename also `Style` to `Flow`?

File: models/pull.go [hunk 32]
```diff
@@ -81,7 +81,7 @@ type PullRequest struct {
 
 	isHeadRepoLoaded bool `xorm:"-"`
 
-	Style PullRequestStyle `xorm:"NOT NULL DEFAULT 0"`
+	Style PullRequestFlow `xorm:"NOT NULL DEFAULT 0"`
```
Comment (lunny) [inline models/pull.go]: As above

File: models/migrations/v189.go (hunk 31 above)
Comment (lunny) [reply models/migrations/v189.go]: @6543 @a1012112796 I mean `Flow PullRequesFlow`.

**Comment by lunny** [comment]
When push to an empty repository, an error returned `remote: error: cannot find hook 'proc-receive'`.

**Comment by lunny** [comment]
Another two problems:
- [ ] The branch link under the title should be disabled because there is no branch created.
- [ ] Documentations should be updated.

**Comment by a1012112796** [comment]
> When push to an empty repository, an error returned `remote: error: cannot find hook 'proc-receive'`.

If a repo created before this pull request, should run ``Resynchronize pre-receive, update and post-receive hooks of all repositories.``. or if create new repo, It will response other error message example view:
Peek 2021-07-25 23-06

 I will  add more check to response a more better error message.

**Comment by lunny** [comment]
Maybe you should rewrite the gitea hooks on migrations if git version matched.

**Comment by a1012112796** [comment]
> Maybe you should rewrite the gitea hooks on migrations if git version matched.

I think just add a note on blog is enough

**Comment by 6543** [comment]
@a1012112796 well since we check for the git version and **assume** if it is high enough we do recive **proc-receive**.
thats why we skip some checks on post - if there are no proc hooks in place, can this be used to pass code into a repo without permissions to do so?

**Comment by a1012112796** [comment]
> @a1012112796 well since we check for the git version and **assume** if it is high enough we do recive **proc-receive**.
> thats why we skip some checks on post - if there are no proc hooks in place, can this be used to pass code into a repo without permissions to do so?

Willn't, Because will check permission in ``pre-receive`` hook.

File: models/migrations/v190.go [hunk 33]
```diff
@@ -0,0 +1,24 @@
...
+
+package migrations
+
+import (
+	"fmt"
+
+	"xorm.io/xorm"
+)
+
+func addAgitStylePullRequest(x *xorm.Engine) error {
```
Comment (lunny) [inline models/migrations/v190.go]: `AgitFlow`, Maybe you could find more places to be changed.

**Review by lunny — APPROVED** [review]
Please change more `PullRequestStyle` to `PullRequestFlow`, otherwise LGTM

**Comment by lunny** [comment]
make L-G-T-M work.

**Comment by wULLSnpAXbWZGYDYyhWTKKspEQoaYxXyhoisqHf** [comment]
🚀

File: services/agit/agit.go [hunk 34]
```diff
@@ -0,0 +1,288 @@
...
+		pr.HeadCommitID = opts.NewCommitIDs[i]
+		if err = pull_service.UpdateRef(pr); err != nil {
+			log.Error("Failed to update pull ref. Error: %v", err)
+			ctx.JSON(http.StatusInternalServerError, map[string]interface{}{
+				"Err": fmt.Sprintf("Failed to update pull ref. Error: %v", err),
+			})
+			return nil
+		}
+
+		pull_service.AddToTaskQueue(pr)
```
Comment (wxiaoguang) [inline services/agit/agit.go:234]: @a1012112796 Do we need to call `pull_service.AddToTaskQueue` (`pull_service.StartPullRequestCheckImmediately` in new code) here?


The PR check would also be added in the post-receive hook?

Comment (wxiaoguang) [reply services/agit/agit.go:234]: @a1012112796 ping , what do you think?
